File: coding/src/com/algos/li01_array/revised/L169_MajorityElement.java
package  com.algos.li01_array.revised;

public class L169_MajorityElement {

	public static void main(String[] args) {
		L169_MajorityElement obj = new L169_MajorityElement();
		
		int[] nums = new int[] {3,2,3};
		System.out.println(obj.majorityElement(nums));
	}

    // Problem: https://leetcode.com/problems/majority-element/
    // Idea: Moore's Voting algorithm
    public int majorityElement(int[] nums) {
        int major = -1, cnt = 0;
        for(int num: nums){
            if(num == major){
                cnt++;
            } else if(cnt == 0){
                major = num;
                cnt++;
            } else{
                cnt--;
            }
        }

        return major;
    }    
    
}

File: dsa/practice/01_arrays/l169_majority_element.py
# L169 - Majority Element
#
# Problem: https://leetcode.com/problems/majority-element/
# Idea: Moore's Voting algorithm
# The majority element is the element that appears more than ⌊n/2⌋ times.
# Time: O(n)
# Space: O(1)


def majority_element(nums):
    candidate = None
    count = 0

    # Phase 1: Find potential candidate
    for num in nums:
        if count == 0:
            candidate = num
            count = 1
        elif num == candidate:
            count += 1
        else:
            count -= 1

    return candidate


# -------------------- TEST --------------------
if __name__ == "__main__":
    nums = [3, 2, 3]
    print(majority_element(nums))
